src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.Websockets.cs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.HttpSys.Internal;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    /// <summary>
+    /// Represents the websocket portion of the <see cref="IISHttpContext"/>
+    /// </summary>
+    internal partial class IISHttpContext
+    {
+        private bool _wasUpgraded; // Used for detecting repeated upgrades in IISHttpContext
+
+        private IISAwaitable _readWebSocketsOperation;
+        private IISAwaitable _writeWebSocketsOperation;
+        private TaskCompletionSource<object> _upgradeTcs;
+
+        private async Task StartBidirectionalStream()
+        {
+            // IIS allows for websocket support and duplex channels only on Win8 and above
+            // This allows us to have two tasks for reading the request and writing the response
+            var readWebsocketTask = ReadWebSockets();
+            var writeWebsocketTask = WriteWebSockets();
+            await Task.WhenAll(readWebsocketTask, writeWebsocketTask);
```

Comment:
You don't need to await here, you can just return the task directly and remove the async from the method definition.